### DISCUSSION TITLE
Soft Body Objects

### DISCUSSION THREAD

**Discussion by milorddev** [enable3d/enable3d#184] [Help / Questions] [answered]
Hey all,

Ive been trying to re-create the soft body example from Ammo.js using enable3d, and I'm having a very hard time doing so. Here's the example for reference: [https://threejs.org/examples/physics_ammo_volume.html](https://threejs.org/examples/physics_ammo_volume.html)

My setup: In a React project, I'm setting up a the `enable3d/phaser-extension` with the `Phaser.WEBGL`, setting, rendering out
```
enable3d(() => new Phaser.Game(this.config)).withPhysics('../assets/ammo')
```

Now I understand that `withPhysics` sets up Ammo behind the scenes, and I can interact with it in the `Scene3D` using `this.physics`, but how can I just import Ammo and do something like `let x = Ammo.btQuaternion(1,1,1,1);` for example?

I just want to mess with soft bodies :D

**Comment by milorddev** [comment]
Oh dang, nevermind. turns out I didnt have to do anything. The `withPhysics` function already makes Ammo global, just make sure to write Ammo stuff like `let x = new Ammo.btQuaternion(1,1,1,1);` and if eslint is yelling at you with a `nodef` message, go into `package.json` and add this:
```
"eslintConfig": {
    "globals": {
      "Ammo": true
    }
  },
```
I'll keep updating this discussion if I get further with the softbody stuff

**Comment by milorddev** [comment]
Alright, next issue. for some reason, I cant seem to access `physicsWorld.getWorldInfo()`. I'm met with a `function does not exist` error. I tried looking into it and finding out examples, however they all pointed to it being within `this.physics.physicsWorld.getWorldInfo()` for the standalone version of enable3d. I am using the `phaser-extension`: trying the above gave me an error, so I modified to `this.third.physics.physicsWorld.getWorldInfo()`, where `physicsWorld` seems to only return `physicsWorld: x {hy: 8771408}` ...

Anyone have insight? @yandeu ? I need this for the `softBodyHelpers.CreatePatch()` function to work.

**Reply by yandeu** [reply]
Have you enabled softBodies?
Somewhere you have to set:
`{ softBodies: true }`

**Reply by milorddev** [reply]
I have not, where would I add that to? I added it to the phaser config
```
this.config = {
      type: Phaser.WEBGL,
      transparent: true,
      scale: {
        mode: Phaser.Scale.FIT,
        autoCenter: Phaser.Scale.CENTER_BOTH,
        width: GAME_WIDTH,
        height: GAME_HEIGHT
      },
      width: GAME_WIDTH,
        height: GAME_HEIGHT,
      parent: "phaser-game",
      scene: [MoonWarsGame],
      softBodies: true,
      ...Canvas()
    };
```
    
I'm pretty sure thats wrong, I noticed on one of the example files for standalone, its supposed to go into the `PhysicsLoader`, what would the equivalent be for `withPhysics()` ?

**Reply by milorddev** [reply]
Ok! Had to track it down but within the `Scene3D` class is a class to `Third` which contains the config that has the softBody property! All you have to do within enable3d is `this.accessThirdDimension({softBodies: true});`

Slowly but surely making some progress! I'll post further updates

**Comment by milorddev** [comment, marked as answer]
Alright! Got softbody physics working in the `phaser-extension`, within a React project with a sphere example! Here's what I did.

First, I init the phaser stuff like usual (Remember this is within a React project, so modify for your instance as you see fit
```
export default class Game extends React.Component {
  componentDidMount() {
    this.config = {
      type: Phaser.WEBGL,
      transparent: true,
      scale: {
        mode: Phaser.Scale.FIT,
        autoCenter: Phaser.Scale.CENTER_BOTH,
        width: GAME_WIDTH,
        height: GAME_HEIGHT
      },
      width: GAME_WIDTH,
        height: GAME_HEIGHT,
      parent: "phaser-game",
      scene: [MoonWarsGame],
      ...Canvas()
    };
  }

  shouldComponentUpdate() {
    return false;
  }

  render() {
    console.log('loaded window')
    enable3d(() => new Phaser.Game(this.config)).withPhysics('../assets/ammo')
    return <div id="phaser-game" />;
  }
}
```

Then in my `Scene3D`, I had to make sure that `softBodies` were set so that Ammo can load the proper physics environment
```
init() {
        this.accessThirdDimension({softBodies: true});
        this.third.physics.debug.enable(); // just to see the debug stuff, not needed for prod
    }
```
Next I essentially copied what the Ammo.js soft body volume example had in their code, making small edits to tie it into enable3d
```
init() {
        const mass = 15;
        const pressure = 161;
        const sphereGeometry = new THREE.SphereGeometry( 1, 32, 32 );  //32 by 32 poly sphere
        sphereGeometry.translate( 0, 3, 0 ); // moved it up a tad just to have it fall down onto the scene
        let player = createSoftVolume( sphereGeometry, mass, pressure); 
        this.scene.add('player', player); // add the softbody to the scene
    }
```

The rest of the functionality is copied from [https://github.com/kripken/ammo.js/blob/main/examples/webgl_demo_softbody_volume/index.html](https://github.com/kripken/ammo.js/blob/main/examples/webgl_demo_softbody_volume/index.html)

Specifically, the functions I used were `createSoftVolume`, `processGeometry`, `mapIndices`,  and `isEqual`. You can import `BufferGeometryUtils` from `import { BufferGeometryUtils } from 'three/examples/jsm/utils/BufferGeometryUtils.js';`

Make sure to go into those functions and change some of the code to tie into `enable3d`. For example
```
physicsWorld.getWorldInfo(),
```
becomes
```
this.third.physics.physicsWorld.getWorldInfo()
```

And thats it! If you want the specific code for these function, I can upload them later. Hope this helps!!

**Reply by yandeu** [reply]
@sciencefreak500 Thanks for sharing :)
